packages/go_router/lib/src/delegate.dart

```diff
@@ -42,10 +43,42 @@ class GoRouterDelegate extends RouterDelegate<RouteMatchList>
   /// Set to true to disable creating history entries on the web.
   final bool routerNeglect;
 
-  final GlobalKey<NavigatorState> _key = GlobalKey<NavigatorState>();
-
-  RouteMatchList _matches = RouteMatchList.empty();
+  RouteMatchList _matchList = RouteMatchList.empty();
   final Map<String, int> _pushCounts = <String, int>{};
+  final RouteConfiguration _configuration;
+
+  @override
+  Future<bool> popRoute() {
+    // Iterate backwards through the RouteMatchList until seeing a GoRoute
+    // with a non-null parentNavigatorKey or a ShellRoute with a non-null parentNavigatorKey
+    // and pop from that Navigator instead of the root.
+    NavigatorState? navigator;
+    final int matchCount = _matchList.matches.length;
+    for (int i = matchCount - 1; i >= 0; i--) {
+      final RouteMatch match = _matchList.matches[i];
+      final RouteBase route = match.route;
+      final GlobalKey<NavigatorState>? key = route.parentNavigatorKey;
+
+      // If this is a ShellRoute, then pop one of the subsequent GoRoutes, if
+      // there are any.
+      if (route is ShellRoute && (matchCount - i) > 2) {
+        // Pop from this navigator.
+        navigator = route.navigatorKey.currentState;
+        break;
+      } else if (key != null) {
+        navigator = key.currentState;
+        break;
+      }
+    }
+
+    navigator ??= navigatorKey.currentState;
+
+    if (navigator == null) {
+      return SynchronousFuture<bool>(false);
+    }
+
+    return navigator.maybePop();
+  }
 
   /// Pushes the given location onto the page stack
   void push(RouteMatch match) {
```

Comment:
how would push work if there is shell route in the middle

Reply:
I will double check, but you can push() a ShellRoute just like any other route, I'm not sure what you mean when you say that there's a shell route in the middle.

One issue where push APIs could be a problem is that there's no way to specify which Navigator push() should push onto - it will always be the navigator configured by the GoRoute's navigatorKey I think, but I can check this too.

Reply:
The last match GoRoute may mean to be embedded in a ShellRoute, but now this logic will just grab the page and push on top of the entire screen.